### ISSUE TITLE
Stream and Collectors are not supported

### ISSUE THREAD

**Issue by yaitskov** [cincheo/jsweet#379]
Hi,

Following code is not translatable:
```
import java.util.stream.Stream;
import java.util.stream.Collectors;
Stream.of("hello", "world").collect(Collectors.toList())
```

[Live sandbox ](http://www.jsweet.org/jsweet-live-sandbox/) fails with:
`Line 26: cannot find name 'java'`

**Comment by renaudpawlak** [comment]
Indeed. Streams are not supported yet. But that would be relatively easy to provide basic stream support. All the info to do so is available here: https://github.com/cincheo/jsweet/blob/master/doc/jsweet-language-specifications.md#extending-the-transpiler

Do not hesitate to PR ;)

**Comment by yaitskov** [comment]
Hi,

I've implemented streams, but it looks like transpiler cannot translate client code using j4ts candy with my patch.

```
    protected abstract <T> Stream<T> stream(Collection<T> c);
    
    public void testStreamCollectList() {
        Object o1 = asList(1, 2, 3);
        Collection<Integer> l = asList(1, 2, 3);
        Object o2 = this.<Integer>stream(l).collect(Collectors.toList());
       
    }
```

Error is

```
2017-10-05 19:19:12.012 INFO  JSweetTranspiler:1403 - bundle.ts(28,34): error TS2346: Supplied parameters do not match any signature of call target.
2017-10-05 19:19:12.012 ERROR output:55 - supplied parameters do not match any signature of call target at /home/user/doc/j4ts/j4ts-test/target/base-test/org/jsweet/BaseJreTest.java(76)

```

I tried to investigate the code, but I cannot find where TS2346 is defined or substring "Supplied parameters do not match any signature of call target"

**Comment by yaitskov** [comment]
Oh, Now I see the error comes from node js typescript tool.

generated typescript code:
```
            let o1 : any = java.util.Arrays.asList(1, 2, 3);
            let l : java.util.Collection<number> = java.util.Arrays.asList(1, 2, 3);
            let o2 : any = <any>(this.stream<number>(l).collect<any, any>(java.util.stream.Collectors.toList<any>()));

```

**Comment by yaitskov** [comment]
Method collect is generated with 1 type argument, while generated caller pass 2 parameters. 

```
collect<R>(supplier?: any, accumulator?: any, combiner?: any): any;
```

Actual java method has 2 type arguments.

**Comment by yaitskov** [comment]
It looks like a bug with overloading interface methods. 
2 methods are combined into one, while overloading is working in a typical class. 

```
        public m$int(n : number) {
            console.info(n);
        }

        public m$java_lang_String(s : string) {
            console.info(s);
        }
```

**Comment by renaudpawlak** [comment]
Umm. Ok I see but I am a little bit lost in the details here ;) Could you please try to write a full minimal example that shows the problem? Does your interface belong to the source code, or is it an external interface?

**Comment by yaitskov** [comment]
I faced the problem trying to use streams. 
Java java.util.stream.Stream interface has overloaded method collect with various type parameters.
I see type script supports usual optional parameters, but it's not true about type parameters.

So while you can collapse usual overloaded methods. this doesn't work for overloaded with various number of type parameters.
 
Minimal interface
```
public interface OverloadInterfaceMethodByType {
    <T> void m(T t);
    <A, B> void m(A a, B b);
}
```
jsweet generates 1 method in the interfaces.

As for me I don't understand great benefit of collapsing at all because transpiler mangles overloaded methods in usual classes anyway.
